### DISCUSSION TITLE
Csla and IgniteUI.Blazor

### DISCUSSION THREAD

**Discussion by Brad-IMS** [MarimerLLC/csla#3328] [Questions] [answered]
After many years successfully using Csla 4.5 with Infragistics Windows Forms and WPF grids, I am currently developing a Blazor WebAssembly app using Csla 6.2.2 and Infragistics' IgniteUI.Blazor IgbGrid.  However, the grid does not seem to want to bind with a Csla collection based on either BusinessBindingListBase or BindingListBase.  Has anyone else encountered this combination of technologies, and, if so, how did you get it to work?

**Comment by kcabral817** [comment, marked as answer]
I've used infragistics in the past but moved to Radzen since their controls are free.  I didn't have problems binding the grid.  Can you provide an example of both the front end and code you're using?

Regards

Kevin
________________________________
From: Brad-IMS ***@***.***>
Sent: Wednesday, May 3, 2023 9:30:45 PM
To: MarimerLLC/csla ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [MarimerLLC/csla] Csla and IgniteUI.Blazor (Discussion #3328)


After many years successfully using Csla 4.5 with Infragistics Windows Forms and WPF grids, I am currently developing a Blazor WebAssembly app using Csla 6.2.2 and Infragistics' IgniteUI.Blazor IgbGrid. However, the grid does not seem to want to bind with a Csla collection based on either BusinessBindingListBase or BindingListBase. Has anyone else encountered this combination of technologies, and, if so, how did you get it to work?

**Reply by Brad-IMS** [reply]
The Radzen grid looks intriguing, but I can't bind to that either.  It seems to default to binding to IQueryable LINQ data.  Did you have to do anything special to bind your Csla collection?

**Comment by kcabral817** [comment]
No.  I bind to VM.Model.attributeName in the control.  I'll send you an example tonight.

Regards

Kevin
________________________________

**Comment by kcabral817** [comment]
<RadzenDataGrid @ref="grid0" ColumnWidth="200px"  AllowFiltering="true" FilterMode="FilterMode.Advanced" AllowPaging="true" AllowSorting="true"
                        Data="@Vm.Model" RowCreate=@GridRowCreate TItem="AbsentReasonCodeInfoPersist">
            <Columns>
                <RadzenDataGridColumn TItem="AbsentReasonCodeInfoPersist" Property="CodeId" Title="Code" Visible="false">
                  <EditTemplate Context="absentReasonCd">
                        <RadzenNumeric style="display: block; width: 100%" @bind-Value="@absentReasonCd.CodeId" Name="CodeId" Visible="false" />
                  </EditTemplate>
                </RadzenDataGridColumn>
               
                <RadzenDataGridColumn TItem="AbsentReasonCodeInfoPersist" Property="Description" Title="Description">
                  <EditTemplate Context="absentReasonCd">
                      <RadzenTextBox style="display: block; width: 100%" @bind-Value="@absentReasonCd.Description" Name="Description" />
                      <RadzenRequiredValidator Component="Description" Text="Description is required" style="position: absolute" />
                  </EditTemplate>
                </RadzenDataGridColumn>
                


        [Inject]
        public ViewModel<AbsentReasonCodeInfoListPersist> Vm { get; set; }

        [Inject]
        public Csla.IDataPortal<AbsentReasonCodeInfoPersist> CodePortal { get; set; }

 protected override async Task OnInitializedAsync()
        {
            try
            {
                if (Vm.Model != null)
                    Vm.Model.Clear();

                await Vm.RefreshAsync(() => CodesPortal.FetchAsync());
            }
            catch (Exception ex)
            {
                NotificationService.Notify(new NotificationMessage
                {
                    Severity = NotificationSeverity.Error,
                    Summary = $"Error",
                    Detail = $"Unable to get Absent Reasons."
                });
            }
        }

**Reply by Brad-IMS** [reply]
Still having trouble getting it to work.  For your List class, are you using BusinessBindingListBase or BusinessListBase?  Also, in your earlier comment, you said you were binding to VM.Model.attributeName, but in your code sample I see Data="@Vm.Model".  Can you elaborate on that?

**Comment by kcabral817** [comment]
Please show me your code.

Regards

Kevin
________________________________

**Comment by Brad-IMS** [comment]
I'm attaching the relevant files, plus some elements extracted from larger files.  I'm normally using ADO.Net, but I've also been testing with an EFCore version.  On the collection class, I've tried both BusinessBindingListBase and BusinessListBase.  I'm having trouble connecting to either the Radzen grid or the Infragistics grid.  I'm able to connect to the Infragistics grid if I convert my Csla collection to a List, but then I have to handle binding manually.  BTW, thanks for your help.

```
builder.Services.AddDbContext<StadisDbContext>(options =>
options.UseSqlServer(builder?.Configuration?.GetConnectionString("DefaultConnection") ?? ""));

builder.Services.AddScoped<StadisDbConnection>(options =>
new StadisDbConnection(builder?.Configuration?.GetConnectionString("DefaultConnection") ?? ""));

builder.Services.AddTransient(typeof(StadisAccess.IPOSDal), typeof(StadisAccess.POSDal));
```

    public class POSEntity
    {
        [Key]
        public string? POSID { get; set; } = null!;
        public string? VendorID { get; set; }
        public string? LocationID { get; set; }
        public string? TranTypeDefault { get; set; }
        [NotMapped]
        public int ReturnCode { get; set; }
        [NotMapped]
        public string? ReturnMessage { get; set; }
    }

    public interface IPOSDal
    {
        POSEntity Get(string id);
        List<POSEntity> Get();
        POSEntity Insert(POSEntity pos);
        POSEntity Update(POSEntity pos);
        bool Delete(string id);
    }

[POS.txt](https://github.com/MarimerLLC/csla/files/11488728/POS.txt)
[POSDal(ADO.Net).txt](https://github.com/MarimerLLC/csla/files/11488729/POSDal.ADO.Net.txt)
[POSDal(EF).txt](https://github.com/MarimerLLC/csla/files/11488730/POSDal.EF.txt)
[POSEdit.txt](https://github.com/MarimerLLC/csla/files/11488731/POSEdit.txt)

**Comment by kcabral817** [comment]
Brad,

I would change private RadzenDataGrid<POS>? rgrid;
To
protected RadzenDataGrid<POS> rgrid;


Then put a breakpoint here:  var temp = vm.Exception;

Check vm.Model for data.

I don’t see anything else other than that.  I’ve used BusinessListBase and DynamicListBase and both bind to the grid..

Regards,

Kevin
831 905-0062

**Comment by Brad-IMS** [comment]
Curiouser and curiouser.  I changed the code as shown below:

```
@code {

    protected RadzenDataGrid<POS>? rgrid;

    protected override async Task OnInitializedAsync()
    {
        try
        {
            if (vm.Model != null)
                vm.Model.Clear();

            await vm.RefreshAsync(() => posListPortal.FetchAsync());
            var temp = vm.Exception;
            string x = vm.Model[0].VendorID;
        }
        catch (Exception ex)
        {

        }
    }

    protected override async Task OnParametersSetAsync()
    {
        await vm.RefreshAsync(() => posListPortal.FetchAsync());
        var temp = vm.Exception;
        string x = vm.Model[1].VendorID;
    }

}
```
With breakpoints after the 'string x' in each:

At end of OnInitializedAsync:
No exceptions or error messages
vm.Model.Count = 7 (correct)
rgrid.Count = 0
rgrid.Data = null
rgrid.Data.Items.Count = 0
x has valid data

At end of OnParametersSetAsync:
No exceptions or error messages
vm.Model.Count = 7 (correct)
rgrid.Count = 1
rgrid.Data = StadisClasses.POSList
rgrid.Data.Items.Count = 7
x has valid data
However, grid does not display.

Now for the weird part.
With breakpoint only after OnParametersSetAsync:

At end of OnParametersSetAsync:
No exceptions or error messages
vm.Model.Count = 7 (correct)
rgrid.Count = 0
rgrid.Data = null
rgrid.Data.Items.Count = 0
x has valid data

So, it looks like it may be a timing issue as far as loading the grid, but that doesn't explain why the grid doesn't display once it has data.

**Comment by kcabral817** [comment]
See if you need to notify the binding engine of the change using:

```
protected override async Task OnInitializedAsync()
        {
try
        {
            if (vm.Model != null)
                vm.Model.Clear();

            await vm.RefreshAsync(() => posListPortal.FetchAsync());
            var temp = vm.Exception;
            string x = vm.Model[0].VendorID;
        }
        catch (Exception ex)
        {

        }

            Vm.ModelPropertyChanged += async (s, e) => await InvokeAsync(() => StateHasChanged());
        } 
```

**Reply by Brad-IMS** [reply]
Got the Radzen grid working.  It was a really stupid error.  I had copied all the columns over from my Infragistics grid, and did Find and Replaces to change the wording to match Radzen, but I didn't notice that, unlike Infragistics, they wrap their columns with a <columns></columns>.  As soon as I put that in it started working.  Thanks again for your help.
